stage1/init/kvm.go

```diff
@@ -40,17 +45,129 @@ func KvmPodToSystemd(p *stage1commontypes.Pod, n *networking.Networking) error {
 		return errwrap.Wrap(errors.New("failed to transform networking to units"), err)
 	}
 
-	// volumes
-	// prepare all applications names to become dependency for mount units
-	// all host-shared folder has to become available before applications starts
-	appNames := []types.ACName{}
-	for _, runtimeApp := range p.Manifest.Apps {
-		appNames = append(appNames, runtimeApp.Name)
+	return nil
+}
+
+func mountSharedVolumes(root string, volumes []types.Volume, ra *schema.RuntimeApp) error {
+	app := ra.App
+	appName := ra.Name
+	vols := make(map[types.ACName]types.Volume)
+	for _, v := range volumes {
+		vols[v.Name] = v
+	}
+
+	sharedVolPath := common.SharedVolumesPath(root)
+	if err := os.MkdirAll(sharedVolPath, stage1initcommon.SharedVolPerm); err != nil {
+		return errwrap.Wrap(errors.New("could not create shared volumes directory"), err)
+	}
+	if err := os.Chmod(sharedVolPath, stage1initcommon.SharedVolPerm); err != nil {
+		return errwrap.Wrap(fmt.Errorf("could not change permissions of %q", sharedVolPath), err)
+	}
+
+	mounts := stage1initcommon.GenerateMounts(ra, vols)
+	for _, m := range mounts {
+		vol := vols[m.Volume]
+
+		if vol.Kind == "empty" {
+			p := filepath.Join(sharedVolPath, vol.Name.String())
+			if err := os.MkdirAll(p, stage1initcommon.SharedVolPerm); err != nil {
+				return errwrap.Wrap(fmt.Errorf("could not create shared volume %q", vol.Name), err)
+			}
+			if err := os.Chown(p, *vol.UID, *vol.GID); err != nil {
+				return errwrap.Wrap(fmt.Errorf("could not change owner of %q", p), err)
+			}
+			mod, err := strconv.ParseUint(*vol.Mode, 8, 32)
+			if err != nil {
+				return errwrap.Wrap(fmt.Errorf("invalid mode %q for volume %q", *vol.Mode, vol.Name), err)
+			}
+			if err := os.Chmod(p, os.FileMode(mod)); err != nil {
+				return errwrap.Wrap(fmt.Errorf("could not change permissions of %q", p), err)
+			}
+		}
+
+		readOnly := stage1initcommon.IsMountReadOnly(vol, app.MountPoints)
+		var source string
+		switch vol.Kind {
+		case "host":
+			source = vol.Source
+		case "empty":
+			source = filepath.Join(common.SharedVolumesPath(root), vol.Name.String())
+		default:
+			return fmt.Errorf(`invalid volume kind %q. Must be one of "host" or "empty"`, vol.Kind)
+		}
+		absAppRootfs, err := filepath.Abs(common.AppRootfsPath(root, appName))
+		if err != nil {
+			return fmt.Errorf(`could not evaluate absolute path for application rootfs in app: %v`, appName)
+		}
+
+		absDestination, err := filepath.Abs(filepath.Join(absAppRootfs, m.Path))
+		if err != nil {
+			return fmt.Errorf(`could not evaluate absolute path for application volume path %q in: %v`, m.Path, appName)
+		}
+		if !strings.HasPrefix(absDestination, absAppRootfs) {
+			return fmt.Errorf("path escapes app's root: %v", absDestination)
+		}
```

Comment:
We were discussing about this earlier in https://github.com/coreos/rkt/pull/2328#discussion_r57722925